core/capabilities/ccip/ccipevm/rmncrypto.go

```diff
@@ -0,0 +1,179 @@
+// nolint:lll
+const encodingUtilsAbiRaw = `[{"inputs":[],"stateMutability":"nonpayable","type":"constructor"},{"inputs":[],"name":"DoNotDeploy","type":"error"},{"inputs":[{"internalType":"bytes32","name":"rmnReportVersion","type":"bytes32"},{"components":[{"internalType":"uint256","name":"destChainId","type":"uint256"},{"internalType":"uint64","name":"destChainSelector","type":"uint64"},{"internalType":"address","name":"rmnRemoteContractAddress","type":"address"},{"internalType":"address","name":"offrampAddress","type":"address"},{"internalType":"bytes32","name":"rmnHomeContractConfigDigest","type":"bytes32"},{"components":[{"internalType":"uint64","name":"sourceChainSelector","type":"uint64"},{"internalType":"bytes","name":"onRampAddress","type":"bytes"},{"internalType":"uint64","name":"minSeqNr","type":"uint64"},{"internalType":"uint64","name":"maxSeqNr","type":"uint64"},{"internalType":"bytes32","name":"merkleRoot","type":"bytes32"}],"internalType":"struct Internal.MerkleRoot[]","name":"destLaneUpdates","type":"tuple[]"}],"internalType":"struct RMNRemote.Report","name":"rmnReport","type":"tuple"}],"name":"_rmnReport","outputs":[],"stateMutability":"nonpayable","type":"function"}]`
+const addressEncodeAbiRaw = `[{"name":"method","type":"function","inputs":[{"name": "", "type": "address"}]}]`
+
+var (
+	EncodingUtilsABI abi.ABI
+	AddressEncodeABI abi.ABI
+)
+
+func init() {
+	var err error
+
+	EncodingUtilsABI, err = abi.JSON(strings.NewReader(encodingUtilsAbiRaw))
+	if err != nil {
+		panic(fmt.Errorf("failed to parse encoding utils ABI: %v", err))
+	}
+
+	AddressEncodeABI, err = abi.JSON(strings.NewReader(addressEncodeAbiRaw))
+	if err != nil {
+		panic(fmt.Errorf("failed to parse address encode ABI: %v", err))
+	}
+}
+
+// EVMRMNCrypto is the RMNCrypto implementation for EVM chains.
+type EVMRMNCrypto struct{}
+
+// Interface compliance check
+var _ cciptypes.RMNCrypto = (*EVMRMNCrypto)(nil)
+
+func NewEVMRMNCrypto() *EVMRMNCrypto {
+	return &EVMRMNCrypto{}
+}
+
+type evmRMNRemoteReport struct {
+	DestChainId                 *big.Int
+	DestChainSelector           uint64
+	RmnRemoteContractAddress    common.Address
+	OfframpAddress              common.Address
+	RmnHomeContractConfigDigest [32]byte
+	DestLaneUpdates             []evmInternalMerkleRoot
+}
+
+type evmInternalMerkleRoot struct {
+	SourceChainSelector uint64
+	OnRampAddress       []byte
+	MinSeqNr            uint64
+	MaxSeqNr            uint64
+	MerkleRoot          [32]byte
+}
+
+func (r *EVMRMNCrypto) VerifyReportSignatures(
+	_ context.Context,
+	sigs []cciptypes.RMNECDSASignature,
+	report cciptypes.RMNReport,
+	signerAddresses []cciptypes.Bytes,
+) error {
+	const v = 27 // used in ecrecover
+
+	if sigs == nil {
+		return fmt.Errorf("no signatures provided")
+	}
+	if report.LaneUpdates == nil {
+		return fmt.Errorf("no lane updates provided")
+	}
+
+	rmnVersionHash := crypto.Keccak256Hash([]byte(report.ReportVersion))
```

Comment:
I wonder if this version hash should be pulled out into a `var` (similar to how we have the LEAF_DOMAIN_SEPARATOR for the message hasher) and have this struct versioned (i.e EVMRMNCryptoV1_6 for V1.6)

Reply:
Hmm could be but then the interface/common-types might also not be compatible with next versions.
I recommend waiting for next version before defining a generic report type to use across different versions.